### PR TITLE
Apply AutoProps to TextBox settings in SUI

### LINKED ISSUE #13827: [Settings]: Screen Reader is not announcing name for 'Tab Title' edit field inside Startup page.
### Windows Terminal version

1.15.2003.0

### Windows build number

10.0.25179.1000

### Other Software

**Test Environment:**
OS: Window 11 22H2 OS Build 25179.1000
App: Windows Terminal Preview
AT: Narrator

### Steps to reproduce

**Repro Steps:**
1. Open Windows Terminal.
2. Open Narrator AT.
3. Inside Settings> Startup page, Expand Tab title control then navigate to Edit field control and observe the issue.

User Experience:
People with Visual impairments like Screen Reader will be impacted here as well as they might not be able to figure out edit field purpose.

Guideline Reference:
https://www.w3.org/WAI/WCAG21/Understanding/name-role-value

Attachments:
[Edit Tab Title.zip](https://github.com/microsoft/terminal/files/9401843/Edit.Tab.Title.zip)

### Expected Behavior

 Screen Reader should announce 'Edit Tab Title' for 'Tab Title' edit field inside Startup page.

### Actual Behavior

 Screen Reader is only announcing 'Edit' for 'Tab Title' edit field inside Startup page.

**Note Issue:** Same issue repro with Edit field which present inside 'Settings> Startup> Word delimiters' and 'Settings> Command Prompt> Name'.

Observation: 
Issue repro with JAWS and NVDA Screen Reader.
Screen Reader users might figure out name by its group control name.

### PR DESCRIPTION
**Pull request by carlos-zamora**
We already were setting the automation properties on the expander, however, we were not setting it on the content when an expander was present. This change applies the automation properties to both the expander and the child content (i.e. TextBox).

Closes #13827

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 1]
```diff
@@ -133,43 +133,46 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
         {
             if (const auto& obj{ content.try_as<DependencyObject>() })
             {
-                base = obj;
+                base.push_back(obj);
             }
         }
 
-        if (base)
+        for (auto obj : base)
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:152]:
```suggestion
        for (const auto& obj : base)
```

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 2]
```diff
@@ -133,43 +133,46 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         _UpdateOverrideSystem();
 
         // Get the correct base to apply automation properties to
-        DependencyObject base{ nullptr };
+        std::vector<DependencyObject> base(2, nullptr);
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:136]: wait, doesn't this create a vector with `nullptr, nullptr` and then you add two _new_ elements to it (`nullptr, nullptr, expander, content`)?

Recommend .. not doing that. I saw the nullptr check down on R154 and it made me realize something was amiss. You're never adding a nullptr, so why check?

**Review by DHowett — CHANGES_REQUESTED** [review]
Wait, won't this make the UIA tree confusing? In every case, we will have two objects with the same automation properties (AND TOOLTIPS) in the tree, one of which is outside of and much larger than the other.

It feels like it'll tell the user, "There's a control here that will let you set the title. Expand. There's a control here that will let you set the title. Text box."

**Comment by DHowett** [comment]
I must admit.. I still don't see how this helps. The TextBox is not a Panel, and so both it and the containing control will receive the same tooltip and the same accessibility description. In addition, the tooltip _is also displayed on the screen inside the expander_.

image

image

None of these things actually seem to be the right place for that tooltip, _or_ the associated automation description.

**Comment by carlos-zamora** [comment]
> I must admit.. I still don't see how this helps. The TextBox is not a Panel, and so both it and the containing control will receive the same tooltip and the same accessibility description. In addition, the tooltip _is also displayed on the screen inside the expander_.
> 
> image
> 
> image
> 
> None of these things actually seem to be the right place for that tooltip, _or_ the associated automation description.

The tooltip is a completely separate thing. I would need to verify it, but I think that's an issue with the expander itself where the tooltip is applied to all the child items by default. See Launch Parameters > "Use system default". All of the launch params have the main tooltip for the expander, but "use system default" gets its own because we applied it.

The help text acts differently though. I added the panel check because for something like "Launch size" that looks like this...
Launch size expander has two child number boxes
the inner stack panel would get the name and help text applied. This caused a weird behavior as follows:
- focus on expander: read expander name and help text
- tab to child control: read expander name and help text (as control type "grouping")
- (I think) then tabbing to another child control would fix itself and actually read the child control's name and help text
By not adding the auto props to the panels, we completely skip the second bullet point above. So it works much better. Another good example to look at here is Defaults > Command line. 

As for something like Interaction > Word Delimiters, I still think this is fine, but I want to verify it with a few screen reader users. Working on sending them a build. The behavior now is...
- on expander: "word delimiters, grouping, <help text>, collapsed"
- invoke expander: "expanded"
- tab to text box: "word delimiters, edit <setting value>, <help text>"
The key words here are "grouping" and "edit". We need the auto props on the expander so that the user knows what to expand/collapse (basically an opt-in for editing the setting). But we also need the same auto props on the text box so that the user knows what setting/field is specifically being edited (and what value is present). Removing one or the other is definitely problematic. And I would even go as far to argue that the name _should_ be the same because you want the association between the expander, the setting, and the text field itself. Even as a sighted user, I'd say the same general flow exists right? It's like this: I want to edit setting X --> read expander --> click it to opt-in to modifying the value. I'd argue the same association exists?

Regardless, again, sending a build to a few people for feedback.

**Comment by carlos-zamora** [comment]
Yup! Verified with JAWS user Alex Benoit. It's 100% ok to have multiple controls have the same name as long as they are exposed as different controls. Since it's being read as "expanded/collapsed" (expander) vs "edit" (text box), he understands the UI tree.

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp [hunk 3]
```diff
@@ -133,43 +133,46 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         _UpdateOverrideSystem();
 
         // Get the correct base to apply automation properties to
-        DependencyObject base{ nullptr };
+        std::vector<DependencyObject> base;
+        base.reserve(2);
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:137]: You can use `til::small_vector<DependencyObject, 2> base;` here and drop the reserve. Or `til::some` (although we might want to remove til::some in favor of the more generic `til::small_vector`).

**Comment by lhecker** [comment]
A potential counter point is that `til::some` vs. `til::small_vector` is a bit like `FAIL_FAST_IF` vs. `assert`. If you do make a mistake handling `til::some` it might crash you executable. Uncaught exceptions in coroutines for instance result in a `std::terminate()` IIRC.

File: src/cascadia/TerminalSettingsEditor/SettingContainer.cpp (hunk 3 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/SettingContainer.cpp:137]: `DependencyObject` doesn't have a default ctor, so neither work :(

**Review by zadjii-msft — APPROVED** [review]
sure, you tested it with someone who actually uses this everyday? that's better than any review I could give.

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
Now, just conflicts to fix! I bet it's "removing the redundant tooltips"

**Pull request merged**